### ISSUE TITLE
Get lines in a range {*x*, *y*} ?

### ISSUE THREAD

**Issue by xged** [Gabriel439/Haskell-Turtle-Library#66]
... where *x*, *y* is a number or a pattern
Eg `sed '0,/pattern/'`

**Comment by Gabriel439** [comment]
Most variations on this can be solved by providing the following utility:

```haskell
keep :: Fold a Bool -> Shell a -> Shell a
```

... that would obey these laws:

```haskell
keep (pure True) = id

keep (liftA2 (&&) x y) = keep x . keep y
```

For example, if you wanted to filter lines with an index greater than 5, you would write:

```haskell
keep (fmap (> 5) Fold.length) :: Shell a -> Shell a
```

For patterns, you could provide two utilities:

```haskell
-- | Returns `True` before the matching the pattern and `False` after
before :: Pattern a -> Fold a Bool

-- | Returns `False` before matching the pattern and `True` after
after :: Pattern a -> Fold a Bool
```

... and then you could write:

```haskell
keep (liftA2 (&&) (after p1) (before p2)) :: Shell Text -> Shell Text
```

Would that solution be alright with you?

**Comment by xged** [comment]
Ok, though at the moment Fold datatype is confusing

**Comment by xged** [comment]
-- remove lines from top to the last occurrence of text
    removeTop :: Text -> Shell Text -> Shell [Text]
    removeTop pat s = tail . last . groupBy (const $ not . isInfixOf pat) <$> fold s Fold.list

**Comment by crocket** [comment]
https://mattyhall.github.io/posts/2015-02-20-scripting-in-haskell.html was written in past, and it suffered the lack of `nl :: Num n => Shell a -> Shell (n, a)` and `filter` for `Shell`.

**Comment by Gabriel439** [comment]
@crocket `Shell` already has `filter` in the form of `Control.Monad.mfilter`, which does the right thing.  I can mention it in the tutorial if that's what you had in mind

**Comment by crocket** [comment]
Yes, it's better to mention `mfilter` in your tutorial because Matthew Hall suffered a perceived lack of `filter`.

**Comment by codygman** [comment]
+1 for mentioning mfilter

**Comment by Gabriel439** [comment]
Alright, this is documented in c045300ce8563534bbeb9fd3b23ea9f5864fe042
